### PR TITLE
rocBLAS: Add an option to disable tensile

### PR DESCRIPTION
**Pull request by teonnik**
- Tensile takes a lot of time to compile, add an option to disable it
- Change variant name to be consistent with rocsolver
- Change default variant value to be consistent with rocsolver

Ideally Tensile should be a separate package but that can be handled in a different PR.

### CODE REVIEW

**Comment by srekolam** [comment]
@cgmb  ,can you please review it

**Comment by cgmb** [comment]
Does rocSOLVER build with `rocblas~tensile`? IIRC, there used to be missing symbols if you tried to do that.

**Comment by teonnik** [comment]
@cgmb 

> Does rocSOLVER build with rocblas~tensile? IIRC, there used to be missing symbols if you tried to do that.

The package builds fine. However, I checked for missing symbols and indeed `gemm`s are missing. I thought that `rocblas` would have those even without `tensile`, I guess I was wrong. :/

**Comment by haampie** [comment]
@cgmb do you think rocblas could have a fallback gemm that's quick to build?

**Comment by cgmb** [comment]
That's been on my personal wishlist for a while... but I have a rather long wishlist. My suggestion would be to open an issue requesting the feature from rocBLAS.

File: var/spack/repos/builtin/packages/rocblas/package.py
Comment (cgmb) [inline var/spack/repos/builtin/packages/rocblas/package.py]: I really don't like defaulting to one random architecture. Yeah, I have gfx906 in my workstation, but now everyone on W6800, MI100, MI200, etc. needs to manually specify the architecture. And they probably will only find out they need to do that by building everything, running their program, having it fail, and debugging what's missing.

**Comment by cgmb** [comment]
I haven't actually tested it, but I like this change aside from the default amdgpu_target.

File: var/spack/repos/builtin/packages/rocblas/package.py
Comment (teonnik) [reply var/spack/repos/builtin/packages/rocblas/package.py]: That is a good point. I chose that architecture to match the one set in `rocsolver` and to reduce the compile time.

Comment (srekolam) [reply var/spack/repos/builtin/packages/rocblas/package.py]: @teonnik  Tensile is already in a separate package called 
rocm-tensile.

Comment (teonnik) [reply var/spack/repos/builtin/packages/rocblas/package.py]: @srekolam Thanks! I missed that one! I will try to rewrite the PR in terms of that package.

Comment (cgmb) [reply var/spack/repos/builtin/packages/rocblas/package.py]: I'm not sure that makes sense. Tensile exists in three forms:
1. It's a command-line executable for generating kernels, benchmarking them, and saving the best solutions in YAML files.
2. It's a build system component that reads YAML solution files, generates the specified kernels, and invokes the compiler to build them. It stores the results of that process in TensileLibrary.dat, and various code object files (`*.co`, `*.hsaco`).
3. It's a run-time library that will load the TensileLibrary.dat and code object files when you call it to do a GEMM, and choose the ideal kernel for your specific parameters.

I could be totally wrong, but I would think rocm-tensile only fulfills role (1). In any case, the YAML solution files are a part of rocBLAS, not Tensile (i.e., they are solutions to rocBLAS's problems). You won't save any time building Tensile separately, because (2) is what takes so long, and it needs rocBLAS's data. So, there's not much benefit to separating them.

I'm also not sure if (3) can be built ahead of time. @lringham, can you use the built version of rocm-tensile to build rocBLAS, or does it _require_ the Tensile source? For reference, [this is how Tensile is being built in the separate package](https://github.com/spack/spack/blob/66c70b82927fc7d3e37009461407fe282873a9a9/var/spack/repos/builtin/packages/rocm-tensile/package.py).

Comment (cgmb) [reply var/spack/repos/builtin/packages/rocblas/package.py]: I forwarded this discussion to one of the senior members of the rocBLAS team, who responded, "if they build rocBLAS and Tensile differently than us, [...] it will be a maintenance problem to support them."

I would not recommend using a separate rocm-tensile.

Comment (teonnik) [reply var/spack/repos/builtin/packages/rocblas/package.py]: @cgmb OK, I will not try to use `rocm-tensile` as a dependency in this package. I will also try to return the previous previous behavior which was to build all architectures by default. I would like to do that without an explicit `all` value but I am not sure if that is possible at the moment. I will have a go at it next week.

Comment (cgmb) [reply var/spack/repos/builtin/packages/rocblas/package.py]: > I will also try to return the previous previous behavior which was to build all architectures by default. I would like to do that without an explicit all value but I am not sure if that is possible at the moment. I will have a go at it next week.

By default, all the ROCm math libraries will build for all the architectures that they support when `-DAMDGPU_TARGETS` is not passed. IMO, until Spack has proper support for detecting and setting GPU architectures, this should be the default.

IMO, don't worry about the 'all' thing in this PR. Just keep the existing behaviour. Then we open another PR that fixes the defaults and adds a consistent 'default' variant for all the ROCm math libraries.

**Comment by cgmb** [comment]
There will be some minor merge conflicts to resolve if #27436 is merged first, but they should be pretty easy to deal with.

File: var/spack/repos/builtin/packages/rocblas/package.py
Comment (cgmb) [inline var/spack/repos/builtin/packages/rocblas/package.py]: I didn't notice this was defaulted to `False` until I went to use it, but it should be `True`. The library won't function as documented without Tensile. It's a feature that exists mostly for specific sorts of testing and debugging.

**Review by cgmb — CHANGES_REQUESTED** [review]
The defaults on the variants are not appropriate. This should not be merged until they're changed.

By the way, there are build speed improvements coming for Tensile in ROCm 5.0/5.1, so you do have something to look forward to.

File: var/spack/repos/builtin/packages/rocblas/package.py
Comment (cgmb) [reply var/spack/repos/builtin/packages/rocblas/package.py]: I opened #28907 for the `AMDGPU_TARGETS` defaults and propagation change.

Comment (teonnik) [reply var/spack/repos/builtin/packages/rocblas/package.py]: @cgmb  I see that you have incorporated most of the changes from this PR. Perhaps it is better to continue on your PR and close this one?

**Comment by teonnik** [comment]
I addressed the comments regarding the default variant and copied the changes for the variant from #28907 for rocblas. This now has the same changes for rocblas as in #28907 except that the default for tensile is set to True. If merging this PR before #28907 is preferable, then this can be merged, if not, then this PR can be closed and only #28907 can be merged (after the default value is changed there).

**Review by cgmb — APPROVED** [review]
This change still needs testing to verify it doesn't break anything, but the code looks good.

**Comment by cgmb** [comment]
> There will be some minor merge conflicts to resolve if #27436 is merged first, but they should be pretty easy to deal with.

It has been merged, so the conflicts will need to be resolved.

**Comment by cgmb** [comment]
This PR was closed because it was subsumed into #28907.

I talked with the rocBLAS team about this and they were receptive to fleshing out the `BUILD_WITH_TENSILE=OFF` variant into something more usable by third-parties. The optimized Tensile kernels are important for achieving peak performance with rocblas, but they do introduce build-time and run-time complexity that's not always needed.

**Pull request closed without merging**
